mp/src/game/client/momentum/ui/HUD/hud_speedometer_label.cpp

```diff
@@ -0,0 +1,285 @@
+#include "cbase.h"
+
+#include "hud_speedometer_label.h"
+
+#include <vgui_controls/AnimationController.h>
+
+#include "hud_comparisons.h"
+#include "iclientmode.h"
+#include "momentum/util/mom_util.h"
+#include "mom_player_shared.h"
+
+#include "tier0/memdbgon.h"
+
+// 1 unit = 19.05mm -> 0.01905m -> 0.00001905Km(/s) -> 0.06858Km(/h)
+#define UPS_TO_KMH_FACTOR 0.06858f
+// 1 unit = 0.75", 1 mile = 63360. 0.75 / 63360 ~~> 0.00001184"(/s) ~~> 0.04262MPH
+#define UPS_TO_MPH_FACTOR 0.04262f
+
+// velocity has to change this much before it is colored as increase/decrease
+#define COLORIZE_DEADZONE 2.0f
+
+using namespace vgui;
+
+SpeedometerLabel::SpeedometerLabel(Panel *parent, const char *panelName, SpeedometerColorize_t colorizeType)
+    : Label(parent, panelName, ""), m_eColorizeType(colorizeType), m_pFlAlpha(nullptr),
+      m_pSzAnimationName(nullptr), m_bSupportsEnergyUnits(false), m_bDoneFading(false),
+      m_eUnitType(SPEEDOMETER_UNITS_UPS), m_pRangeList(nullptr), m_bDrawComparison(true), 
+      m_bSupportsSeparateComparison(true), m_cvarGravity("sv_gravity")
+{
+    CreateComparisonLabel();
+    Reset();
+}
+
+void SpeedometerLabel::SetVisible(bool bVisible)
+{
+    BaseClass::SetVisible(bVisible);
+    // parent's layout depends on the visiblity of this, so invalidate it
+    GetParent()->InvalidateLayout();
+}
+
+// creates a separate label for the comparison & pins it to this
+void SpeedometerLabel::CreateComparisonLabel()
+{
+    char name[BUFSIZELOCL];
+    Q_snprintf(name, sizeof(name), "%sComparison", GetName());
+    m_pComparisonLabel = new Label(GetParent(), name, "");
+    m_pComparisonLabel->SetName(name);
+    m_pComparisonLabel->SetPos(0, 0);
+    m_pComparisonLabel->PinToSibling(GetName(), PIN_TOPLEFT, PIN_TOPRIGHT);
+    m_pComparisonLabel->SetFont(GetFont());
+    m_pComparisonLabel->SetAutoTall(true);
+    m_pComparisonLabel->SetAutoWide(true);
+    m_pComparisonLabel->SetContentAlignment(a_center);
+    m_pComparisonLabel->SetPinCorner(PIN_TOPLEFT, 0, 0);
+}
```

Comment:
This can actually move to the constructor, it's small enough and is actually constructing vital things for the class.